kv/txn_coord_sender_test.go

```diff
@@ -317,15 +318,59 @@ func TestTxnCoordSenderEndTxn(t *testing.T) {
 	s := createTestDB(t)
 	defer s.Stop()
 
-	key := roachpb.Key("a")
-	txn := client.NewTxn(*s.DB)
-	if err := txn.Put(key, []byte("value")); err != nil {
-		t.Fatal(err)
-	}
-	if err := txn.Commit(); err != nil {
-		t.Fatal(err)
+	// 4 cases: no deadline, past deadline, equal deadline, future deadline.
+	for i := 0; i < 4; i++ {
+		key := roachpb.Key("key: " + strconv.Itoa(i))
+		txn := client.NewTxn(*s.DB)
+		// Initialize the transaction
+		if err := txn.Put(key, []byte("value")); err != nil {
+			t.Fatal(err)
+		}
+
+		var deadline *roachpb.Timestamp
+		switch i {
+		case 0:
+			// No deadline.
+		case 1:
+			// Past deadline.
+			ts := txn.Proto.Timestamp.Prev()
+			deadline = &ts
+		case 2:
+			// Equal deadline.
+			deadline = &txn.Proto.Timestamp
+		case 3:
+			// Future deadline.
+			ts := txn.Proto.Timestamp.Next()
+			deadline = &ts
+		}
+
+		{
+			err := txn.CommitBy(deadline)
+			switch i {
+			case 0:
+				// No deadline.
+				if err != nil {
+					t.Error(err)
+				}
+			case 1:
+				// Past deadline.
+				if err != nil {
+					t.Error(err)
+				}
+			case 2:
+				// Equal deadline.
+				if err != nil {
+					t.Error(err)
+				}
+			case 3:
+				// Future deadline.
```

Comment:
Isn't this backwards? Future deadlines should allow for a commit and past deadlines should fail.